File: 2025/9/9_28_leetcode_bc/238.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    let zero = false;
    let total_product = 1;
    let output = [];

    for (const val of nums) {
        if (val === 0) {
            if (zero) {
                total_product = 0;
                break;
            }
            else {
                zero = true;
            }
        } else {
            total_product *= val;
        }
    }

    for (const val of nums) {
        if (val === 0) {
            output.push(total_product);
        } else if (zero) {
            output.push(0);
        } else {
            output.push(Math.floor(total_product / val))
        }
    }

    return output
};

File: 2025/9/9_28_leetcode_bc/238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        zero = False
        total_product = 1
        output = []

        for i in nums:
            if i == 0:
                if zero:
                    total_product = 0
                    break
                zero = True
                continue
            total_product *= i

        for i in nums:
            if i == 0:
                output.append(total_product)
            elif zero:
                output.append(0)
            else:
                output.append(total_product // i)

        return output



